src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+        }
+    }
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_100: [ channel_internal_close shall call srw_lock_release_exclusive. ]*/
+    srw_lock_release_exclusive(channel_internal_ptr->lock);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CHANNEL_INTERNAL), channel_internal_create_and_open, THANDLE(THREADPOOL), threadpool)
+{
+    THANDLE(CHANNEL_INTERNAL) result = NULL;
+
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_098: [ channel_create shall call srw_lock_create. ]*/
+    SRW_LOCK_HANDLE lock = srw_lock_create(false, "channel");
+    if (lock == NULL)
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+        LogError("Failure in srw_lock_create(false, \"channel\")");
+    }
+    else
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_078: [ channel_create shall create a CHANNEL_INTERNAL object by calling THANDLE_MALLOC with channel_internal_dispose as dispose.]*/
+        THANDLE(CHANNEL_INTERNAL) channel_internal = THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose);
+        if (channel_internal == NULL)
+        {
+            /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+            LogError("Failure in THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose)");
+        }
+        else
+        {
+            CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_internal);
+
+            channel_internal_ptr->lock = lock;
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_080: [ channel_create shall store given threadpool in the created CHANNEL_INTERNAL. ]*/
+            THANDLE_INITIALIZE(THREADPOOL)(&channel_internal_ptr->threadpool, threadpool);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_084: [ channel_create shall call DList_InitializeListHead. ]*/
+            DList_InitializeListHead(&channel_internal_ptr->op_list);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_086: [ channel_create shall succeed and return the created THANDLE(CHANNEL). ]*/
+            THANDLE_INITIALIZE_MOVE(CHANNEL_INTERNAL)(&result, &channel_internal);
+            goto all_ok;
+        }
+        srw_lock_destroy(lock);
+    }
+all_ok:
+    return result;
+}
+
+static void execute_callbacks(void* context)
+{
+    if (context == NULL)
+    {
+        LogError("Invalid arguments: void* context=%p", context);
+    }
+    else
+    {
+        CHANNEL_OP* channel_op = (CHANNEL_OP*)context;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_145: [ execute_callbacks shall call the stored callback(s) with the result of the operation. ]*/
+        if (channel_op->pull_callback)
+        {
+            channel_op->pull_callback(channel_op->pull_context, channel_op->result, channel_op->data);
+        }
+        if (channel_op->push_callback)
+        {
+            channel_op->push_callback(channel_op->push_context, channel_op->result);
+        }
```

Comment:
why are both callbacks served from the same thread? 

Wouldn't it make more sense that that pull callback and the push callbacks can be served at the same time?

Reply:
We want push callback to execute after pull callback has succeeded.